File: 1051-height-checker/1051.py
class Solution:
    def heightChecker(self, heights: List[int]) -> int:
        # Approach 2: Counting Sort
        #               T.C: O(N), S.C: O(1)
        count = [0] * 101
        for h in heights:
            count[h] += 1

        res = 0
        curr_height = 1
        for h in heights:
            while count[curr_height] == 0:
                curr_height += 1
            if h != curr_height:
                res += 1
            count[curr_height] -= 1
        return res

        """
        # Appraoch 1: Sorting
        #           T.C: O(n log n), S.C: O(n)
        sorted_heights = sorted(heights)

        res = 0
        for i in range(0, len(heights)):
            if sorted_heights[i] != heights[i]:
                res += 1
        return res
        """

File: 1051-height-checker/1051.cpp
/*
 * @lc app=leetcode id=1051 lang=cpp
 *
 * [1051] Height Checker
 */

// @lc code=start
const static auto fast = [] {
  ios_base::sync_with_stdio(false);
  cin.tie(nullptr);
  cout.tie(nullptr);
  return 0;
}();

class Solution {
public:
  int heightChecker(vector<int>& heights) {
    /* Approach:
        1. Counting Sort -- T.C: O(N), S.C: O(N)
            - Use Counting Sort to simulate the sorted order without actually sorting, then count
       mismatches with the original array.
    */

    int n          = heights.size();
    int count[101] = {0};

    // Count frequency of each height
    for (int h : heights) {
      count[h]++;
    }

    int res         = 0;
    int curr_height = 1;
    // Starting from smallest possible height sort without sorting :D
    for (int h : heights) {
      // Skip until non-zero count
      while (count[curr_height] == 0) curr_height++;

      if (h != curr_height) res++;
      count[curr_height]--;
    }

    return res;
  }
};
// @lc code=end
